test-data/unit/check-overloading.test

```diff
@@ -2937,3 +2937,322 @@ def add_proxy(x, y):
 tup = (1, '2')
 reveal_type(foo(lambda (x, y): add_proxy(x, y), tup))  # E: Revealed type is 'builtins.str*'
 [builtins fixtures/primitives.pyi]
+
+[case testOverloadWithClassMethods]
+from typing import overload
+
+class Wrapper:
+    @overload
+    @classmethod
+    def foo(cls, x: int) -> int: ...
+    @overload
+    @classmethod
+    def foo(cls, x: str) -> str: ...
+    @classmethod
+    def foo(cls, x): pass
+
+reveal_type(Wrapper.foo(3))      # E: Revealed type is 'builtins.int'
+reveal_type(Wrapper.foo("foo"))  # E: Revealed type is 'builtins.str'
+
+[builtins fixtures/classmethod.pyi]
+
+[case testOverloadWithInconsistentClassMethods]
+from typing import overload
+
+class Wrapper1:
+    @overload   # E: Overload does not consistently use the "@classmethod" decorator on all function signatures.
+    @classmethod
+    def foo(cls, x: int) -> int: ...
+    @overload
+    @classmethod
+    def foo(cls, x: str) -> str: ...
+    def foo(cls, x): pass
+
+class Wrapper2:
+    @overload   # E: Overload does not consistently use the "@classmethod" decorator on all function signatures.
+    @classmethod
+    def foo(cls, x: int) -> int: ...
+    @overload
+    def foo(cls, x: str) -> str: ...
+    @classmethod
+    def foo(cls, x): pass
+
+class Wrapper3:
+    @overload   # E: Overload does not consistently use the "@classmethod" decorator on all function signatures.
+    def foo(cls, x: int) -> int: ...
+    @overload
+    def foo(cls, x: str) -> str: ...
+    @classmethod
+    def foo(cls, x): pass
+
+[builtins fixtures/classmethod.pyi]
+
+[case testOverloadWithSwappedDecorators]
+from typing import overload
+
+class Wrapper1:
+    @classmethod
+    @overload
+    def foo(cls, x: int) -> int: ...
+
+    @classmethod
+    @overload
+    def foo(cls, x: str) -> str: ...
+
+    @classmethod
+    def foo(cls, x): pass
+
+class Wrapper2:
+    @classmethod
+    @overload
+    def foo(cls, x: int) -> int: ...
+
+    @overload
+    @classmethod
+    def foo(cls, x: str) -> str: ...
+
+    @classmethod
+    def foo(cls, x): pass
+
+class Wrapper3:
+    @classmethod   # E: Overload does not consistently use the "@classmethod" decorator on all function signatures.
+    @overload
+    def foo(cls, x: int) -> int: ...
+
+    @overload
+    def foo(cls, x: str) -> str: ...
+
+    def foo(cls, x): pass
+
+reveal_type(Wrapper1.foo(3))  # E: Revealed type is 'builtins.int'
+reveal_type(Wrapper2.foo(3))  # E: Revealed type is 'builtins.int'
+
+[builtins fixtures/classmethod.pyi]
+
+[case testOverloadFaultyClassMethodInheritance]
+from typing import overload
+
+class A: pass
+class B(A): pass
+class C(B): pass
+
+class Parent:
+    @overload
+    @classmethod
+    def foo(cls, x: B) -> int: ...
+
+    @overload
+    @classmethod
+    def foo(cls, x: str) -> str: ...
+
+    @classmethod
+    def foo(cls, x): pass
+
+class BadChild(Parent):
+    @overload  # E: Signature of "foo" incompatible with supertype "Parent"
```

Comment:
Maybe add few more tests to check overriding an overloaded instance method, with overloaded class methods and vice versa?

Reply:
Ah whoops, I completely forgot to handle this case. Fixed.